### ISSUE TITLE
Very slow 3D CNN on TPU with torch/xla

### ISSUE THREAD

**Issue by gregfriedland** [pytorch/xla#2523]
## ❓ Questions and Help

Hello,
This 3D CNN code has been running well on GPUs such as NVIDIA V100 but this is my first time running this code on a TPU, so I could be missing something obvious. I changed the device type from `cuda` (or `cpu`) to `xm.xla_device()` and my iterations are now taking 1min each with 90% of that time (as reported by cProfile) spent in ~100 calls/iter to `torch.tensor.split`. FWIW, if I run with device=`cpu`, the training runs 10x faster.

I have the debug_run.py logs but I can't upload them in the open here for privacy reasons. I will try to find a way to get them to you.

Could this be because the data is not all the same size in this part of the code? The bottleneck here appears to be in a data transformation step in the main thread  that happens before the data is shown to the model. This data transformation is usually run with a custom CUDA kernel but since that's not possible here, I'm using an alternate implementation which uses `torch` operations such as `tensor.split`.

Best,
Greg

**Comment by taylanbil** [comment]
may I suggest using gist.github.com to get the debug run logs?

**Comment by jysohn23** [comment]
Hi @gregfriedland based on the debug information you shared, it looks like both the inputs and model are dynamic. To use XLA:TPU efficiently you do want to make sure that your model graph (inputs, outputs, and intermediate tensors) are all static. Could you start off by trying to make the inputs static (by padding if that works for you).

Also, it looks like the debug info indicates you didn't run for a full step. Full step would be running the model until the point [`xm.mark_step()`](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L711) is called. Depending on how you're using `torch_xla` in your script, if you're already using [`ParallelLoader`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/parallel_loader.py#L44) or [`MpDeviceLoader`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/parallel_loader.py#L178) it will already implicitly be calling [`mark_step`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/parallel_loader.py#L37). However, if you're not using the above DataLoader wrappers, you'll want to call either [`xm.optimizer_step(optimizer, barrier=True)`](https://github.com/pytorch/xla/blob/master/torch_xla/core/xla_model.py#L755) or call `mark_step()` yourself.
